src/Symfony/Bundle/SecurityBundle/Templating/Helper/LogoutUrlHelper.php

```diff
@@ -97,6 +111,14 @@ public function getLogoutUrl($key)
      */
     private function generateLogoutUrl($key, $referenceType)
     {
+        // Fetch the current provider key from token, if possible
+        if (null === $key && $this->tokenStorage) {
+            $token = $this->tokenStorage->getToken();
+            if ($token && method_exists($token, 'getProviderKey')) {
+                $key = $token->getProviderKey();
+            }
+        }
+
         if (!array_key_exists($key, $this->listeners)) {
             throw new \InvalidArgumentException(sprintf('No LogoutListener found for firewall key "%s".', $key));
```

Comment:
Would it be better to throw a specific exception when key is null? Probably in the condition above when we are not able to find the corresponding key? Something like "Unable to find the current firewall LogoutListener, please specify the provider key."